cumulus/pallets/xcmp-queue/src/lib.rs

```diff
@@ -1168,13 +907,24 @@ impl<T: Config> SendXcm for Pallet<T> {
 
 	fn deliver((id, xcm): (ParaId, VersionedXcm<()>)) -> Result<XcmHash, SendError> {
 		let hash = xcm.using_encoded(sp_io::hashing::blake2_256);
+		debug_assert!(validate_xcm_nesting(&xcm).is_ok(), "Tickets are valid; qed");
 
 		match Self::send_fragment(id, XcmpMessageFormat::ConcatenatedVersionedXcm, xcm) {
 			Ok(_) => {
 				Self::deposit_event(Event::XcmpMessageSent { message_hash: hash });
 				Ok(hash)
 			},
-			Err(e) => Err(SendError::Transport(<&'static str>::from(e))),
+			Err(e) => Err(SendError::Transport(e.into())),
 		}
 	}
 }
+
+/// Checks that the XCM is decodable with `MAX_XCM_DECODE_DEPTH`.
+///
+/// Note that this uses the limit of the sender - not the receiver. It it best effort.
+pub(crate) fn validate_xcm_nesting(xcm: &VersionedXcm<()>) -> Result<(), ()> {
+	xcm.using_encoded(|mut enc| {
+		VersionedXcm::<()>::decode_all_with_depth_limit(MAX_XCM_DECODE_DEPTH, &mut enc).map(|_| ())
+	})
+	.map_err(|_| ())
```

Comment:
Yea it is wasteful. I talked about this with Gav and he commented that this would be the easiest way to do it.  
Otherwise i would need to write a recursive descender that counts the depth. Although even after doing this, it does not actually guarantee that it decodes since the decode function could change.  

The long-term idea here is to swap to double-encoded messages such that each message has its length prefixed. In that manner we can just skip undecodable messages on the receiver side without breaking the decode stream.

Reply:
> The long-term idea here is to swap to double-encoded messages such that each message has its length prefixed. In that manner we can just skip undecodable messages on the receiver side without breaking the decode stream.

Ok, sounds good !